### PR TITLE
feat: batch transfer memberships

### PR DESCRIPTION
**Pull request by DanThomp507**
As part of the work for [[User Page] Batch Transfer Membership](https://trello.com/c/dNTwbp2u/1009-user-page-batch-transfer-membership), we need to create batch transfer membership functionality within the SDK.

**Constraints**:
- The selected pods MUST be on the same controller version. 
- You can't transfer membership and a manager at the same time. You must either do one or the other, but not both. 

**Challenges**:
- Batch transfer functionality for managers is currently out of scope. This requires changes at the smart contracts level, which we're not currently working on or ready to support. 

**Solution**:
- I created a `batchTransferMembership` function to support the transferral of memberships, but not managers. 
- This function checks to ensure all pods are on the same version. 
- It also checks to ensure that the signer is a member of all pods, as well as if the signer is the same as the `fromAddress`. 
- I also added unit tests to cover these scenarios!

### CODE REVIEW

**Comment by DanThomp507** [comment]
[[SDK] Batch Transfer Membership](https://trello.com/c/Cmh49p0J/1096-sdk-batch-transfer-membership)

File: src/batch.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  toAddress: string,
+  podIds: number[],
+  signer?: ethers.Signer,
+): Promise<ethers.providers.TransactionResponse | { to: string; data: string }> {
+  // we need to get the pod object for the podIds for the member and version checks
+  const pods = await Promise.all(podIds.map(async podId => new Pod(podId)));
+  // check the fromAddress
+  const checkedFrom = checkAddress(fromAddress);
+  // check the toAddress
+  const checkedTo = checkAddress(toAddress);
```
Comment (jtlacci) [inline src/batch.ts:32]: we also need to add a check to make sure the `toAddress` isn't already a member of all the pods

File: src/lib/utils.ts [hunk 2]
```diff
@@ -111,3 +112,34 @@ export async function getPreviousModule(safe, module, newController?: string) {
...
+ * @param signer
+ * @returns
+ */
+export function getContract(contractName: string, signer) {
+  const contractJson = getDeployment(contractName, config.network);
+  if (!contractJson) throw new RangeError(`Contract ABI could not be found for ${contractName}`);
+  return new ethers.Contract(contractJson.address, contractJson.abi, signer);
+}
+
+export async function getIsLatestVersion(signer, pod) {
```
Comment (jtlacci) [inline src/lib/utils.ts:128]: are we using this anywhere?

Comment (DanThomp507) [reply src/lib/utils.ts:128]: Nope, we're not! Will remove.

**Review by kevinweaver — COMMENTED** [review]
Partial review, I'll finish tomorrow

File: src/batch.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  // check to see if all pods have the same controller version
+  const isSameVersion = getIsSameVersion(pods.map(pod => pod.controller));
+  if (!isSameVersion) {
+    throw new Error('All pods must be on the same controller version');
+  }
+
+  try {
+    const MemberToken = getMetropolisContract('MemberToken', signer);
+    // this has to be the same length as the podIds array
+    const valueArray = new Array(podIds.length).fill(1);
```
[the remark this answers is not in the document]
Comment (DanThomp507) [reply src/batch.ts:64]: 😂

File: src/batch.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  const checkedTo = checkAddress(toAddress);
+
+  // helper function to see if an address is or isn't already a pod member
+  const isPodMember = async address => Promise.all(pods.map(async pod => pod.isMember(address)));
+
+  // check to see if the toAddress is already a pod member
+  const isToAddressPodMember = await isPodMember(checkedTo);
+
+  // if any value is true, throw an error
+  if (isToAddressPodMember.includes(true)) {
```
[the remark this answers is not in the document]
Comment (DanThomp507) [reply src/batch.ts:41]: yeah that makes complete sense, I was going back and forth on that too. Definitely felt a bit off. Better off keeping it as a single true or false, will update now!

File: src/batch.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+): Promise<ethers.providers.TransactionResponse | { to: string; data: string }> {
+  // we need to get the pod object for the podIds for the member and version checks
+  const pods = await Promise.all(podIds.map(async podId => new Pod(podId)));
+  // check the fromAddress
+  const checkedFrom = checkAddress(fromAddress);
+  // check the toAddress
+  const checkedTo = checkAddress(toAddress);
+
+  // helper function to see if an address is or isn't already a pod member
+  const isPodMember = async address => Promise.all(pods.map(async pod => pod.isMember(address)));
```
[the remark this answers is not in the document]
Comment (DanThomp507) [reply src/batch.ts:35]: I think that makes sense! I wasn't sure if we'd be using it elsewhere but tbh it's better to have this in utils now so we'd have the ability to do so.

**Comment by DanThomp507** [comment]
@kevinweaver addressed those initial comments!

File: test/batch.test.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    await expect(
+      batchTransferMembership(
+        metropolis1WithAdminPod?.members[3],
+        userAddress3,
+        podIds,
+        mockSigner,
+      ),
+    ).rejects.toThrow('Signer did not match the from address');
+  });
+});
```
Comment (kevinweaver) [inline test/batch.test.ts:119]: 👍 I appreciate the thorough testing!

**Review by kevinweaver — APPROVED** [review]
Looks great!

**Pull request merged**